File: internal/pkg/fswatcher/watch_plan.go
package fswatcher

import (
	"fmt"
	"path"
	"path/filepath"
	"slices"
	"strings"

	"github.com/bmatcuk/doublestar/v4"
	"github.com/vormadev/vorma/kit/fsutil"
	"github.com/vormadev/vorma/kit/globset"
	"github.com/vormadev/vorma/kit/set"
)

type watch_plan struct {
	match_set *globset.Set
	rules     globset.Rules
	roots     []watch_root
}

type watch_root struct {
	path    string
	dynamic bool
}

type pattern_prefix_state struct {
	pattern_index int
	prefix_index  int
}

func new_watch_plan(raw_patterns []string) (*watch_plan, error) {
	patterns := make([]string, len(raw_patterns))
	for i, raw_pattern := range raw_patterns {
		patterns[i] = strings.TrimSpace(raw_pattern)
	}
	if len(patterns) == 0 {
		patterns = []string{"."}
	}

	match_set, err := globset.Compile(patterns)
	if err != nil {
		return nil, fmt.Errorf("compile watch patterns: %w", err)
	}

	rules := match_set.Rules()
	roots := set.New[watch_root]()
	for _, rule := range rules {
		if rule.Excluded {
			continue
		}
		roots.Add(watch_root_for_rule(rule))
	}

	root_slice := roots.Slice()
	slices.SortFunc(root_slice, func(a, b watch_root) int {
		if a.path != b.path {
			return strings.Compare(a.path, b.path)
		}
		if a.dynamic == b.dynamic {
			return 0
		}
		if a.dynamic {
			return 1
		}
		return -1
	})

	return &watch_plan{
		match_set: match_set,
		rules:     rules,
		roots:     root_slice,
	}, nil
}

func (p *watch_plan) should_emit(path string, is_dir bool) bool {
	rel_path := watch_rel_path(path)
	if is_dir {
		rel_path += "/"
	}
	return p.match_set.Match(rel_path)
}

func (p *watch_plan) should_watch_dir(path string) bool {
	rel_path := watch_rel_path(path)
	return p.could_match_in_subtree(rel_path)
}

func (p *watch_plan) could_match_in_subtree(dir string) bool {
	dir = normalize_watch_path(dir)

	last_cover_index := -1
	last_cover_excluded := false
	for i, rule := range p.rules {
		if rule_covers_subtree(rule, dir) {
			last_cover_index = i
			last_cover_excluded = rule.Excluded
		}
	}

	if last_cover_index >= 0 && last_cover_excluded {
		for _, rule := range p.rules[last_cover_index+1:] {
			if !rule.Excluded && rule_can_match_in_subtree(rule, dir) {
				return true
			}
		}
		return false
	}

	for _, rule := range p.rules {
		if !rule.Excluded && rule_can_match_in_subtree(rule, dir) {
			return true
		}
	}
	return false
}

func watch_root_for_rule(rule globset.Rule) watch_root {
	if rule.Pattern == "." {
		return watch_root{path: "."}
	}

	pattern_segments := split_watch_path(rule.Pattern)
	literal_segments := make([]string, 0, len(pattern_segments))
	for _, segment := range pattern_segments {
		if segment_has_glob(segment) {
			break
		}
		literal_segments = append(literal_segments, segment)
	}

	if len(literal_segments) == 0 {
		return watch_root{path: "."}
	}

	root_path := path.Join(literal_segments...)
	if len(literal_segments) == len(pattern_segments) {
		return watch_root{path: root_path, dynamic: true}
	}
	return watch_root{path: root_path}
}

func watch_rel_path(path string) string {
	path = fsutil.SysNorm(path)
	rel_path, err := filepath.Rel(".", path)
	if err != nil {
		return normalize_watch_path(path)
	}
	return normalize_watch_path(rel_path)
}

func normalize_watch_path(raw_path string) string {
	raw_path = strings.TrimSpace(raw_path)
	raw_path = filepath.ToSlash(raw_path)
	raw_path = strings.TrimSuffix(raw_path, "/")
	raw_path = path.Clean(raw_path)
	raw_path = strings.TrimPrefix(raw_path, "./")
	if raw_path == "" {
		return "."
	}
	return raw_path
}

func split_watch_path(path string) []string {
	path = normalize_watch_path(path)
	if path == "." {
		return nil
	}
	return strings.Split(path, "/")
}

func segment_has_glob(segment string) bool {
	return strings.ContainsAny(segment, "*?[{")
}

func rule_can_match_in_subtree(rule globset.Rule, dir string) bool {
	if rule.Pattern == "." {
		return true
	}

	dir_segments := split_watch_path(dir)
	if pattern_can_match_prefix(rule.Pattern, dir_segments) {
		return true
	}
	if rule.ChildPattern != "" &&
		pattern_can_match_prefix(rule.ChildPattern, dir_segments) {
		return true
	}
	if rule.DirOnly {
		child_pattern := path.Join(rule.Pattern, "**")
		return pattern_can_match_prefix(child_pattern, dir_segments)
	}
	return false
}

func rule_covers_subtree(rule globset.Rule, dir string) bool {
	if rule.Pattern == "." {
		return true
	}

	for _, ancestor := range watch_path_ancestors(dir) {
		if pattern_matches_path(rule.Pattern, ancestor) &&
			(rule.ChildPattern != "" || rule.DirOnly) {
			return true
		}
		if pattern_suffix_covers_subtree(rule.Pattern, ancestor) {
			return true
		}
	}
	return false
}

func watch_path_ancestors(path string) []string {
	path = normalize_watch_path(path)
	ancestors := []string{path}
	for path != "." {
		path = normalize_watch_path(filepath.Dir(path))
		ancestors = append(ancestors, path)
	}
	return ancestors
}

func pattern_suffix_covers_subtree(pattern string, dir string) bool {
	base, ok := strings.CutSuffix(pattern, "/**")
	if !ok {
		return false
	}
	if base == "" {
		return true
	}
	return pattern_matches_path(base, dir)
}

func pattern_matches_path(pattern string, path string) bool {
	pattern = normalize_watch_path(pattern)
	path = normalize_watch_path(path)
	if pattern == "." {
		return true
	}
	ok, _ := doublestar.Match(pattern, path)
	return ok
}

func pattern_can_match_prefix(pattern string, prefix_segments []string) bool {
	pattern_segments := split_watch_path(pattern)
	seen := map[pattern_prefix_state]bool{}

	var walk func(pattern_index, prefix_index int) bool
	walk = func(pattern_index, prefix_index int) bool {
		if prefix_index == len(prefix_segments) {
			return true
		}
		if pattern_index == len(pattern_segments) {
			return false
		}

		state := pattern_prefix_state{
			pattern_index: pattern_index,
			prefix_index:  prefix_index,
		}
		if seen[state] {
			return false
		}
		seen[state] = true

		pattern_segment := pattern_segments[pattern_index]
		if pattern_segment == "**" {
			if walk(pattern_index+1, prefix_index) {
				return true
			}
			return walk(pattern_index, prefix_index+1)
		}

		ok, _ := doublestar.Match(pattern_segment, prefix_segments[prefix_index])
		if !ok {
			return false
		}
		return walk(pattern_index+1, prefix_index+1)
	}

	return walk(0, 0)
}

File: crates/vorma-build/src/watch_plan.rs
use std::collections::BTreeSet;
use std::path::Path;

use path_clean::PathClean;
use path_slash::PathBufExt;

use crate::config::{relative_path, sys_norm};
use crate::globset;

#[derive(Clone, Debug)]
pub(crate) struct WatchPlan {
	match_set: globset::Set,
	rules: Vec<globset::Rule>,
	pub(crate) roots: Vec<WatchRoot>,
}

#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
pub(crate) struct WatchRoot {
	pub(crate) path: String,
	pub(crate) dynamic: bool,
}

impl WatchPlan {
	pub(crate) fn new(raw_patterns: &[String]) -> Result<Self, String> {
		let mut patterns = raw_patterns
			.iter()
			.map(|pattern| pattern.trim().to_owned())
			.collect::<Vec<_>>();
		if patterns.is_empty() {
			patterns.push(".".to_owned());
		}

		let match_set = globset::compile(&patterns)?;
		let rules = match_set.rules().to_vec();
		let mut roots = BTreeSet::new();
		for rule in &rules {
			if !rule.excluded {
				roots.insert(watch_root_for_rule(rule));
			}
		}

		Ok(Self {
			match_set,
			rules,
			roots: roots.into_iter().collect(),
		})
	}

	pub(crate) fn should_emit(&self, path: &str, is_dir: bool) -> bool {
		let mut rel_path = watch_rel_path(path);
		if is_dir {
			rel_path.push('/');
		}
		self.match_set.is_match(&rel_path)
	}

	pub(crate) fn should_watch_dir(&self, path: &str) -> bool {
		self.could_match_in_subtree(&watch_rel_path(path))
	}

	fn could_match_in_subtree(&self, dir: &str) -> bool {
		let dir = normalize_watch_path(dir);
		let mut last_cover_index = None;
		let mut last_cover_excluded = false;
		for (i, rule) in self.rules.iter().enumerate() {
			if rule_covers_subtree(rule, &dir) {
				last_cover_index = Some(i);
				last_cover_excluded = rule.excluded;
			}
		}

		if let Some(last_cover_index) = last_cover_index
			&& last_cover_excluded
		{
			return self.rules[last_cover_index + 1..]
				.iter()
				.any(|rule| !rule.excluded && rule_can_match_in_subtree(rule, &dir));
		}

		self.rules
			.iter()
			.any(|rule| !rule.excluded && rule_can_match_in_subtree(rule, &dir))
	}
}

fn watch_root_for_rule(rule: &globset::Rule) -> WatchRoot {
	if rule.pattern == "." {
		return WatchRoot {
			path: ".".to_owned(),
			dynamic: false,
		};
	}

	let mut literal_segments = Vec::new();
	let pattern_segments = split_watch_path(&rule.pattern);
	for segment in &pattern_segments {
		if segment_has_glob(segment) {
			break;
		}
		literal_segments.push(segment.clone());
	}

	if literal_segments.is_empty() {
		return WatchRoot {
			path: ".".to_owned(),
			dynamic: false,
		};
	}

	let root_path = literal_segments.join("/");
	WatchRoot {
		dynamic: literal_segments.len() == pattern_segments.len(),
		path: root_path,
	}
}

fn watch_rel_path(path: &str) -> String {
	let path = sys_norm(path);
	relative_path(".", &path)
		.map(|path| normalize_watch_path(path.to_string_lossy()))
		.unwrap_or_else(|| normalize_watch_path(path))
}

fn normalize_watch_path(raw_path: impl AsRef<str>) -> String {
	let raw_path = raw_path.as_ref().trim();
	let mut path = std::path::PathBuf::from_slash(raw_path)
		.clean()
		.to_slash_lossy()
		.trim_end_matches('/')
		.trim_start_matches("./")
		.to_owned();
	if path.is_empty() {
		path = ".".to_owned();
	}
	path
}

pub(crate) fn split_watch_path(path: &str) -> Vec<String> {
	let path = normalize_watch_path(path);
	if path == "." {
		return Vec::new();
	}
	path.split('/').map(str::to_owned).collect()
}

fn segment_has_glob(segment: &str) -> bool {
	segment.contains(['*', '?', '[', '{'])
}

fn rule_can_match_in_subtree(rule: &globset::Rule, dir: &str) -> bool {
	if rule.pattern == "." {
		return true;
	}

	let dir_segments = split_watch_path(dir);
	if pattern_can_match_prefix(&rule.pattern, &dir_segments) {
		return true;
	}
	if !rule.child_pattern.is_empty()
		&& pattern_can_match_prefix(&rule.child_pattern, &dir_segments)
	{
		return true;
	}
	if rule.dir_only {
		let child_pattern = format!("{}/**", rule.pattern);
		return pattern_can_match_prefix(&child_pattern, &dir_segments);
	}
	false
}

fn rule_covers_subtree(rule: &globset::Rule, dir: &str) -> bool {
	if rule.pattern == "." {
		return true;
	}

	for ancestor in watch_path_ancestors(dir) {
		if pattern_matches_path(&rule.pattern, &ancestor)
			&& (!rule.child_pattern.is_empty() || rule.dir_only)
		{
			return true;
		}
		if pattern_suffix_covers_subtree(&rule.pattern, &ancestor) {
			return true;
		}
	}
	false
}

fn watch_path_ancestors(path: &str) -> Vec<String> {
	let mut path = normalize_watch_path(path);
	let mut ancestors = vec![path.clone()];
	while path != "." {
		path = normalize_watch_path(
			Path::new(&path)
				.parent()
				.map(|path| path.to_string_lossy().into_owned())
				.unwrap_or_else(|| ".".to_owned()),
		);
		ancestors.push(path.clone());
	}
	ancestors
}

fn pattern_suffix_covers_subtree(pattern: &str, dir: &str) -> bool {
	let Some(base) = pattern.strip_suffix("/**") else {
		return false;
	};
	base.is_empty() || pattern_matches_path(base, dir)
}

fn pattern_matches_path(pattern: &str, path: &str) -> bool {
	let pattern = normalize_watch_path(pattern);
	let path = normalize_watch_path(path);
	pattern == "." || globset::path_match_unvalidated(&pattern, &path)
}

fn pattern_can_match_prefix(pattern: &str, prefix_segments: &[String]) -> bool {
	let pattern_segments = split_watch_path(pattern);
	let mut seen = BTreeSet::new();

	fn walk(
		pattern_segments: &[String],
		prefix_segments: &[String],
		pattern_index: usize,
		prefix_index: usize,
		seen: &mut BTreeSet<(usize, usize)>,
	) -> bool {
		if prefix_index == prefix_segments.len() {
			return true;
		}
		if pattern_index == pattern_segments.len() {
			return false;
		}
		if !seen.insert((pattern_index, prefix_index)) {
			return false;
		}

		let pattern_segment = &pattern_segments[pattern_index];
		if pattern_segment == "**" {
			if walk(
				pattern_segments,
				prefix_segments,
				pattern_index + 1,
				prefix_index,
				seen,
			) {
				return true;
			}
			return walk(
				pattern_segments,
				prefix_segments,
				pattern_index,
				prefix_index + 1,
				seen,
			);
		}

		if !globset::path_match_unvalidated(pattern_segment, &prefix_segments[prefix_index]) {
			return false;
		}
		walk(
			pattern_segments,
			prefix_segments,
			pattern_index + 1,
			prefix_index + 1,
			seen,
		)
	}

	walk(&pattern_segments, prefix_segments, 0, 0, &mut seen)
}

#[cfg(test)]
mod tests {
	use super::*;

	#[test]
	fn watch_plan_uses_dot_root_for_unanchored_globs_and_literal_roots_for_subtrees() {
		let plan = WatchPlan::new(&[
			"src/**/*.rs".to_owned(),
			"public".to_owned(),
			"!target/**".to_owned(),
		])
		.unwrap();

		assert_eq!(
			plan.roots,
			vec![
				WatchRoot {
					path: ".".to_owned(),
					dynamic: false,
				},
				WatchRoot {
					path: "src".to_owned(),
					dynamic: false,
				},
			],
		);
	}

	#[test]
	fn watch_plan_understands_excluded_subtrees_with_later_includes() {
		let plan = WatchPlan::new(&[
			".".to_owned(),
			"!src/generated/**".to_owned(),
			"src/generated/keep.rs".to_owned(),
		])
		.unwrap();

		assert!(plan.should_watch_dir("src"));
		assert!(plan.should_watch_dir("src/generated"));
		assert!(plan.should_emit("src/generated/keep.rs", false));
		assert!(!plan.should_emit("src/generated/drop.rs", false));
	}
}
